sale_coupon_advanced/tests/test_sale_coupon_forced_reward_product.py

```diff
@@ -0,0 +1,147 @@
+            self.program_forced.id,
+            programs.ids,
+            "`sale_coupon_program._filter_programs_from_common_rules` should "
+            "return the program for the forced reward product",
+        )
+
+    def test_02_get_reward_line_values__result(self):
+        """
+            Testing `sale_order._get_reward_line_values` strict formalism return
+
+            IMPORTANT : be aware of the partner's lang as the discount product
+            name is changed towards reward product name (addition of
+            translatable `Free product - ` string)
+        """
+        taxes = [(4, tax.id, False) for tax in self.product_B.taxes_id]
+        if self.order_forced.fiscal_position_id:
+            taxes = self.order_forced.fiscal_position_id.map_tax(taxes)
+        self.reward_values = [
+            {
+                "sequence": 998,
+                "name": self.reward_product.name,
+                "product_id": self.reward_product.id,
+                "price_unit": self.reward_product.lst_price,
+                "is_reward_line": False,
+                "product_uom_qty": self.program_forced.reward_product_quantity,
+                "product_uom": self.discount_product.uom_id.id,
+                "tax_id": taxes,
+            },
+            {
+                "sequence": 999,
+                "name": self.discount_product.name,
+                "product_id": self.discount_product.id,
+                "price_unit": -self.reward_product.lst_price,
+                "is_reward_line": True,
+                "product_uom_qty": self.program_forced.reward_product_quantity,
+                "product_uom": self.reward_product.uom_id.id,
+                "tax_id": taxes,
+            },
+        ]
+
+        # Sorting by sequences as the interface does
+        reward_values = sorted(
+            self.order_forced._get_reward_line_values(self.program_forced),
+            key=lambda k: k["sequence"],
+        )
+        self.assertEquals(
+            self.reward_values,
+            reward_values,
+            "`sale_order._get_reward_line_values` should return forced reward "
+            "values",
+        )
+
+    def test_03_order_line__forced_reward_result(self):
+        """
+            Testing global result in sale_order.order_line :
+                - all lines quantity must be 3
+                - reward and discount respectively based on their product id
+                - product discount should canceled the product reward price
+        """
+        self.order_forced.recompute_coupon_lines()
+        reward_and_discount_ids = [self.reward_product.id, self.discount_product.id]
+        all_lines = self.order_forced.order_line
+        reward_and_discount_line = all_lines.filtered(
+            lambda line: line.product_id.id in reward_and_discount_ids
+        )
+        only_discount_line = reward_and_discount_line.filtered(
+            lambda line: line.price_unit == -self.reward_product.lst_price
+        )
```

Comment:
As an additional check, `sale.order.line` have an attribute named `is_reward_line`. Please make sure that each line is a reward line.
```python
self.assertTrue(all(line.is_reward_line for line in only_discount_line))
```

Reply:
Can you say it back please ? I don't get the point.

Reply:
Is it very considered as a discount line ? Check that the `sale_order_line` attribute attribute `is_reward_line` is actually set to `True`.
The code is here, you just have to add it.

Reply:
Fixed